File: adt2keyb/go32/keyboard_init.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

void keyboard_init (void) {
  __dpmi_paddr p;

  DBG_ENTER ("keyboard_init");

  memset (keydown, false, sizeof (keydown));
  SysKeyboardCallback = SysKeyboardUserProc;

  lock_SysKeyboardIRQ (true);
  lock_SysKeyboardUserProc (true);

  __dpmi_get_protected_mode_interrupt_vector (9, &SysKeyboardOldIRQ);

  p.offset32 = (uint32_t) SysKeyboardIRQ;
  p.selector = _go32_my_cs ();
  __dpmi_set_protected_mode_interrupt_vector (9, &p);

  DBG_LEAVE (); //keyboard_init
}

File: adt2keyb/go32/pas/keyboard_init.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure keyboard_init;
begin
  _dbg_enter ({$I %FILE%}, 'keyboard_init');

  FillChar(keydown,SizeOf(keydown),FALSE);
  user_proc_ptr := @int09_user_proc;
  lock_data(user_proc_ptr,SizeOf(user_proc_ptr));
  lock_data(DosMemSelector,SizeOf(DosMemSelector));
  lock_code(@int09_user_proc,DWORD(@int09_user_proc_end)-DWORD(@int09_user_proc));
  lock_code(@newint09_proc,DWORD(@newint09_proc_end)-DWORD(@newint09_proc));
  newint09_handler.offset := @newint09_proc;
  newint09_handler.segment := get_cs;
  get_pm_interrupt($09,oldint09_handler);
  set_pm_interrupt($09,newint09_handler);

  _dbg_leave; //EXIT //keyboard_init
end;
